File: backend/app/api/v1/endpoints/comptabilite.py
from datetime import date
from uuid import UUID

from fastapi import APIRouter, Depends, Query, status
from fastapi.responses import Response
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.deps import require_any_permission, require_permission
from app.models.user import User
from app.schemas.comptabilite import (
    BudgetLigneCreate,
    BudgetSuiviResponse,
    CategorieDepenseResponse,
    CompteTresorerieResponse,
    DepenseCreate,
    DepenseRefus,
    DepenseResponse,
    EcritureResponse,
    RapportFinancierResponse,
    TresorerieResponse,
)
from app.services import comptabilite_service

router = APIRouter()

READ_PERMISSIONS = ("expenses.create", "expenses.validate", "reports.view")
CREATE_PERMISSION = "expenses.create"
VALIDATE_PERMISSION = "expenses.validate"


@router.get("/categories", response_model=list[CategorieDepenseResponse])
async def list_categories(
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_any_permission(*READ_PERMISSIONS)),
):
    return await comptabilite_service.list_categories(db)


@router.get("/comptes", response_model=list[CompteTresorerieResponse])
async def list_comptes(
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_any_permission(*READ_PERMISSIONS)),
):
    return await comptabilite_service.list_comptes_avec_solde(db)


@router.get("/tresorerie", response_model=TresorerieResponse)
async def tresorerie(
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_any_permission(*READ_PERMISSIONS)),
):
    return await comptabilite_service.get_tresorerie(db)


@router.get("/depenses", response_model=list[DepenseResponse])
async def list_depenses(
    statut: str | None = Query(default=None),
    annee_scolaire_id: UUID | None = Query(default=None),
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_any_permission(*READ_PERMISSIONS)),
):
    return await comptabilite_service.list_depenses(db, statut, annee_scolaire_id)


@router.post("/depenses", response_model=DepenseResponse, status_code=status.HTTP_201_CREATED)
async def create_depense(
    data: DepenseCreate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(require_permission(CREATE_PERMISSION)),
):
    return await comptabilite_service.create_depense(db, data, user.id)


@router.post("/depenses/{depense_id}/soumettre", response_model=DepenseResponse)
async def soumettre_depense(
    depense_id: UUID,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_permission(CREATE_PERMISSION)),
):
    return await comptabilite_service.soumettre_depense(db, depense_id)


@router.post("/depenses/{depense_id}/valider", response_model=DepenseResponse)
async def valider_depense(
    depense_id: UUID,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(require_permission(VALIDATE_PERMISSION)),
):
    return await comptabilite_service.valider_depense(db, depense_id, user.id)


@router.post("/depenses/{depense_id}/refuser", response_model=DepenseResponse)
async def refuser_depense(
    depense_id: UUID,
    data: DepenseRefus,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(require_permission(VALIDATE_PERMISSION)),
):
    return await comptabilite_service.refuser_depense(db, depense_id, data, user.id)


@router.post("/budget", status_code=status.HTTP_201_CREATED)
async def upsert_budget(
    data: BudgetLigneCreate,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_permission(VALIDATE_PERMISSION)),
):
    ligne = await comptabilite_service.upsert_budget_ligne(db, data)
    return {"id": str(ligne.id), "montant_prevu": str(ligne.montant_prevu)}


@router.get("/budget/suivi", response_model=BudgetSuiviResponse)
async def budget_suivi(
    annee_scolaire_id: UUID = Query(...),
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_any_permission(*READ_PERMISSIONS)),
):
    return await comptabilite_service.get_budget_suivi(db, annee_scolaire_id)


@router.get("/journal", response_model=list[EcritureResponse])
async def journal(
    date_debut: date | None = Query(default=None),
    date_fin: date | None = Query(default=None),
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_any_permission(*READ_PERMISSIONS)),
):
    return await comptabilite_service.list_journal(db, date_debut, date_fin)


@router.get("/rapports/financier", response_model=RapportFinancierResponse)
async def rapport_financier(
    date_debut: date = Query(...),
    date_fin: date = Query(...),
    annee_scolaire_id: UUID | None = Query(default=None),
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_any_permission(*READ_PERMISSIONS)),
):
    return await comptabilite_service.get_rapport_financier(db, date_debut, date_fin, annee_scolaire_id)


@router.get("/export/csv")
async def export_csv(
    date_debut: date = Query(...),
    date_fin: date = Query(...),
    annee_scolaire_id: UUID | None = Query(default=None),
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_any_permission(*READ_PERMISSIONS)),
):
    rapport = await comptabilite_service.get_rapport_financier(db, date_debut, date_fin, annee_scolaire_id)
    journal = await comptabilite_service.list_journal(db, date_debut, date_fin)
    content = comptabilite_service.export_rapport_csv(rapport, journal)
    return Response(content=content, media_type="text/csv", headers={
        "Content-Disposition": 'attachment; filename="rapport_financier.csv"',
    })


@router.get("/export/excel")
async def export_excel(
    date_debut: date = Query(...),
    date_fin: date = Query(...),
    annee_scolaire_id: UUID | None = Query(default=None),
    db: AsyncSession = Depends(get_db),
    _: User = Depends(require_any_permission(*READ_PERMISSIONS)),
):
    content = await comptabilite_service.export_excel_rapport(db, date_debut, date_fin, annee_scolaire_id)
    return Response(content=content, media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet", headers={
        "Content-Disposition": 'attachment; filename="rapport_financier.xlsx"',
    })

File: frontend/src/lib/api/comptabilite.ts
import type {
  BudgetSuivi,
  CategorieDepense,
  CompteTresorerie,
  Depense,
  Ecriture,
  RapportFinancier,
  Tresorerie,
} from "@/types/comptabilite";
import { apiDownload, apiFetch } from "./client";

const base = "/api/v1/comptabilite";

export async function listCategoriesDepense(token: string) {
  return apiFetch<CategorieDepense[]>(`${base}/categories`, {}, token);
}

export async function listComptesTresorerie(token: string) {
  return apiFetch<CompteTresorerie[]>(`${base}/comptes`, {}, token);
}

export async function getTresorerie(token: string) {
  return apiFetch<Tresorerie>(`${base}/tresorerie`, {}, token);
}

export async function listDepenses(token: string, statut?: string, anneeScolaireId?: string) {
  const params = new URLSearchParams();
  if (statut) params.set("statut", statut);
  if (anneeScolaireId) params.set("annee_scolaire_id", anneeScolaireId);
  const q = params.toString();
  return apiFetch<Depense[]>(`${base}/depenses${q ? `?${q}` : ""}`, {}, token);
}

export async function createDepense(
  token: string,
  data: {
    categorie_id: string;
    libelle: string;
    montant: number;
    date_depense: string;
    compte_tresorerie_id: string;
    annee_scolaire_id?: string;
    reference_piece?: string;
  },
) {
  return apiFetch<Depense>(`${base}/depenses`, { method: "POST", body: JSON.stringify(data) }, token);
}

export async function soumettreDepense(token: string, depenseId: string) {
  return apiFetch<Depense>(`${base}/depenses/${depenseId}/soumettre`, { method: "POST" }, token);
}

export async function validerDepense(token: string, depenseId: string) {
  return apiFetch<Depense>(`${base}/depenses/${depenseId}/valider`, { method: "POST" }, token);
}

export async function refuserDepense(token: string, depenseId: string, motif: string) {
  return apiFetch<Depense>(
    `${base}/depenses/${depenseId}/refuser`,
    { method: "POST", body: JSON.stringify({ motif }) },
    token,
  );
}

export async function getBudgetSuivi(token: string, anneeScolaireId: string) {
  return apiFetch<BudgetSuivi>(`${base}/budget/suivi?annee_scolaire_id=${anneeScolaireId}`, {}, token);
}

export async function listJournal(token: string, dateDebut: string, dateFin: string) {
  return apiFetch<Ecriture[]>(
    `${base}/journal?date_debut=${dateDebut}&date_fin=${dateFin}`,
    {},
    token,
  );
}

export async function getRapportFinancier(
  token: string,
  dateDebut: string,
  dateFin: string,
  anneeScolaireId?: string,
) {
  const params = new URLSearchParams({ date_debut: dateDebut, date_fin: dateFin });
  if (anneeScolaireId) params.set("annee_scolaire_id", anneeScolaireId);
  return apiFetch<RapportFinancier>(`${base}/rapports/financier?${params}`, {}, token);
}

export async function downloadRapportCsv(
  token: string,
  dateDebut: string,
  dateFin: string,
  anneeScolaireId?: string,
) {
  const params = new URLSearchParams({ date_debut: dateDebut, date_fin: dateFin });
  if (anneeScolaireId) params.set("annee_scolaire_id", anneeScolaireId);
  await apiDownload(`${base}/export/csv?${params}`, token, "rapport_financier.csv");
}

export async function downloadRapportExcel(
  token: string,
  dateDebut: string,
  dateFin: string,
  anneeScolaireId?: string,
) {
  const params = new URLSearchParams({ date_debut: dateDebut, date_fin: dateFin });
  if (anneeScolaireId) params.set("annee_scolaire_id", anneeScolaireId);
  await apiDownload(`${base}/export/excel?${params}`, token, "rapport_financier.xlsx");
}
